[i2s_audio] Use high-pass filter for dc offset correction

doobry opened the pull request:
Change exponential moving average for a high pass filter to correct dc offset based around 16khz sample rate of PDM.

Should give a faster response and work for both negative and positive offsets.

Tested with an M5 Atom Echo with PDM mic where existing solution did not correct dc offset.

# What does this implement/fix?

Existing solution did not work on an M5 Atom Echo with a negative offset.  I believe using a high-pass filter is a more 'standard' way of correcting dc offsets, so have implemented the simplest one I could.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
microphone:
  - id: echo_microphone
    platform: i2s_audio
    adc_type: external
    i2s_din_pin: GPIO23
    pdm: true
    channel: stereo
    correct_dc_offset: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

doobry committed:
[i2s_audio] Use high-pass filter for dc offset correction

Change exponential moving average for a high pass filter to correct dc offset based around 16khz.

Should give a faster response and work for both negative and positive offsets.

Tested with an M5 Atom Echo with PDM mic where existing solution did not correct dc offset.

doobry committed:
[i2s_audio] Use better documented dc offset filter

From https://www.musicdsp.org/en/latest/Filters/135-dc-filter.html:

    y(n) = x(n) - x(n-1) + R * y(n-1)
    R = 1 - (pi * 2 * frequency / samplerate)

Calculate R value for 40Hz on a 16kHz sample rate:
    R = 1 - (pi * 2 * 40 / 16000)
    R = 0.9842920367

Transform floating point to bit-shifting approximation:
    output = input - prev_input + R * prev_output
    output = input - prev_input + (prev_output - (prev_output >> S))

Approximate bit-shift value from R:
    R = 1 - (1 >> S)
    R = 1 - 2^-S
    0.9842920367 = 1 - 2^-S
    S ≈ 6

stas-sl commented:
I wonder how previous implementation was supposed to work for offsets < 1000. As `DC_OFFSET_MOVING_AVERAGE_COEFFICIENT_DENOMINATOR = 1000` and if you divide `new_offset` by 1000 using integers, there will be always 0 for `new_offset < 1000`. Or +/- 1000 offsets are considered small, so they could be ignored?

```cpp
dc_offset_ = new_offset / DC_OFFSET_MOVING_AVERAGE_COEFFICIENT_DENOMINATOR 
            + (DC_OFFSET_MOVING_AVERAGE_COEFFICIENT_DENOMINATOR - 1) * this->dc_offset_ 
            /  DC_OFFSET_MOVING_AVERAGE_COEFFICIENT_DENOMINATOR;
```
As for your implementation, you mention that R depends on sample rate, which could be configured in config and for 48KHz, according to your formula to get similar effect, R would need be `0.99947640122`, but you are using a fixed `R = 0.9842920367`. Or it will be still good enough for the task?

doobry commented:
The ballance with choosing R for the purpose of DC blocking is the time it takes to respond vs how much low frequencies are attenuated.   I believe that the code was originally introduced to fix DC offset on PDM based mics that are hardcoded to run at 16kHz so that is what I used to get a 'reasonable' R for the purpose which is then converted to an approximate value for the bitshift.

I don't know what effect the previous code would have on the audio signal, but I believe it is very slow to respond as it only adapts once per block of samples, instead of continously like the proposed code and kind of of has an R of 1- 1/1000 = 0.999.  I have not profiled it, and have not found any references dicussing it being used it as a method for dc offset correction, but I assume it will introduce small discreet jumps in offset between each sample block as it tries to correct the offset.  So it will have an effect on the audio but less than this proposed change due to the longer response time.

I am not an expert in singal processing, I just searched around for a solution to the issue I was seeing, and this kind of filtering appears to be pretty standard.

At 48khz I think the frequency would be about 120Hz, which might be too much for high fidelity audio folks, but then I'm not sure would they be needing to fix a DC offset.  We could choose a higher R closer to 1, it will still block DC but take longer to ramp up.

If you want a longer read this article has a dicussion of the same algorithm and an implementation using floats instead of the transform to bit shifting I did.
https://kleinembedded.com/quick-and-simple-digital-filters-for-smoothing-and-dc-removal/

doobry committed:
[i2s_audio] Change cutoff frequency of dc offset correction

From https://www.musicdsp.org/en/latest/Filters/135-dc-filter.html:

    y(n) = x(n) - x(n-1) + R * y(n-1)
    R = 1 - (pi * 2 * frequency / samplerate)

From https://en.wikipedia.org/wiki/Hearing_range:
    The human range is commonly given as 20Hz up.

From https://en.wikipedia.org/wiki/High-resolution_audio:
    A reasonable upper bound for sample rate seems to be 96kHz.

Calculate R value for 20Hz on a 96kHz sample rate:
    R = 1 - (pi * 2 * 20 / 96000)
    R = 0.9986910031

Transform floating point to bit-shifting approximation:
    output = input - prev_input + R * prev_output
    output = input - prev_input + (prev_output - (prev_output >> S))

Approximate bit-shift value S from R:
    R = 1 - (1 >> S)
    R = 1 - (1 / 2^S)
    R = 1 - 2^-S
    0.9986910031 = 1 - 2^-S
    S = 9.57732 ~= 10

Actual R from S:
    R = 1 - 2^-10 = 0.9990234375

Confirm this has effect outside human hearing on 96000kHz sample:
    0.9990234375 = 1 - (pi * 2 * f / 96000)
    f = 14.9208Hz

Confirm this has effect outside human hearing on PDM 16kHz sample:
    0.9990234375 = 1 - (pi * 2 * f / 16000)
    f = 2.4868Hz

doobry commented:
I have recordings from debugging the assitant pipeline, which is how I knew I had a problem, and that this fixes it.

But they are only when it is trigged, not from the moment the microphone is turned on and this code would start working.  You will only see the stabaisation on the waveform in the first 320ms, and that is what I don't know how to record.

Thanks for the hint with the streaming audio.

kahrendt reviewed (approved):
Great work, very appreciated it!
